File: chazm-api/src/main/java/runtimemodels/chazm/api/relation/manager/NeedsManager.java
package runtimemodels.chazm.api.relation.manager;

import runtimemodels.chazm.api.entity.Attribute;
import runtimemodels.chazm.api.entity.Role;
import runtimemodels.chazm.api.id.UniqueId;
import runtimemodels.chazm.api.relation.Needs;

import java.util.Set;

/**
 * The {@linkplain NeedsManager} interface defines the API for managing {@linkplain Needs}.
 *
 * @author Christopher Zhong
 * @since 7.0.0
 */
public interface NeedsManager {
    /**
     * Creates a needs relation between a {@linkplain Role} and an {@linkplain Attribute} in this {@linkplain NeedsManager}.
     *
     * @param roleId      the {@linkplain UniqueId} that represents the {@linkplain Role}.
     * @param attributeId the {@linkplain UniqueId} that represents the {@linkplain Attribute}.
     */
    void addNeeds(UniqueId<Role> roleId, UniqueId<Attribute> attributeId);

    /**
     * Returns a set of {@linkplain Attribute}s that is needed by a {@linkplain Role} in this {@linkplain NeedsManager}.
     *
     * @param id the {@linkplain UniqueId} that represents the {@linkplain Role}.
     * @return a set of {@linkplain Attribute}s.
     */
    Set<Attribute> getNeeds(UniqueId<Role> id);

    /**
     * Returns a set of {@linkplain Role}s that needs an {@linkplain Attribute} in this {@linkplain NeedsManager}.
     *
     * @param id the {@linkplain UniqueId} that represents the {@linkplain Attribute}.
     * @return a set of {@linkplain Role}s.
     */
    Set<Role> getNeededBy(UniqueId<Attribute> id);

    /**
     * Removes a needs relation between a {@linkplain Role} and an {@linkplain Attribute} from this {@linkplain NeedsManager}.
     *
     * @param roleId      the {@linkplain UniqueId} that represents the {@linkplain Role}.
     * @param attributeId the {@linkplain UniqueId} that represents the {@linkplain Attribute}.
     */
    void removeNeeds(UniqueId<Role> roleId, UniqueId<Attribute> attributeId);

    /**
     * Removes all needs relations from this {@linkplain NeedsManager}.
     */
    void removeAllNeeds();

}

File: chazm-api/src/main/java/runtimemodels/chazm/api/relation/manager/NeedsManager.kt
package runtimemodels.chazm.api.relation.manager

import runtimemodels.chazm.api.entity.Attribute
import runtimemodels.chazm.api.entity.Role
import runtimemodels.chazm.api.id.AttributeId
import runtimemodels.chazm.api.id.RoleId
import runtimemodels.chazm.api.id.UniqueId
import runtimemodels.chazm.api.relation.Needs

/**
 * The [NeedsManager] interface defines the API for managing [Needs].
 *
 * @author Christopher Zhong
 * @since 7.0.0
 */
interface NeedsManager {
    /**
     * Creates a needs relation between a [Role] and an [Attribute] in this [NeedsManager].
     *
     * @param roleId      the [UniqueId] that represents the [Role].
     * @param attributeId the [UniqueId] that represents the [Attribute].
     */
    fun addNeeds(roleId: RoleId, attributeId: AttributeId)

    /**
     * Returns a set of [Attribute]s that is needed by a [Role] in this [NeedsManager].
     *
     * @param id the [UniqueId] that represents the [Role].
     * @return a set of [Attribute]s.
     */
    fun getNeeds(id: RoleId): Set<Attribute>

    /**
     * Returns a set of [Role]s that needs an [Attribute] in this [NeedsManager].
     *
     * @param id the [UniqueId] that represents the [Attribute].
     * @return a set of [Role]s.
     */
    fun getNeededBy(id: AttributeId): Set<Role>

    /**
     * Removes a needs relation between a [Role] and an [Attribute] from this [NeedsManager].
     *
     * @param roleId      the [UniqueId] that represents the [Role].
     * @param attributeId the [UniqueId] that represents the [Attribute].
     */
    fun removeNeeds(roleId: RoleId, attributeId: AttributeId)

    /**
     * Removes all needs relations from this [NeedsManager].
     */
    fun removeAllNeeds()

}
